src/ucp/core/ucp_ep.c

```diff
@@ -139,36 +149,61 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
     ucs_debug("created ep %p to %s %s", ep, ucp_ep_peer_name(ep), message);
     return UCS_OK;
 
+err_free_keys:
+    ucs_free(ucp_ep_ext_gen(ep)->keys);
 err_free_ep:
     ucs_strided_alloc_put(&worker->ep_alloc, ep);
 err:
     return status;
 }
 
-ucs_status_t ucp_worker_create_ep(ucp_worker_h worker, const char *peer_name,
-                                  const char *message, ucp_ep_h *ep_p)
+ucs_status_t ucp_worker_create_ep(ucp_worker_h worker, unsigned ep_init_flags,
+                                  const char *peer_name, const char *message,
+                                  ucp_ep_h *ep_p)
 {
     ucs_status_t status;
     ucp_ep_h ep;
 
     status = ucp_ep_create_base(worker, peer_name, message, &ep);
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
     ucs_list_add_tail(&worker->all_eps, &ucp_ep_ext_gen(ep)->ep_list);
 
+    ucs_assert(ucp_ep_ext_gen(ep)->keys->local == UCP_WORKER_PTR_KEY_INVALID);
+    status = ucs_ptr_map_put(&worker->ptr_map, ep,
```

Comment:
why do we still need to add to the map if UCP_EP_INIT_ERR_MODE_PEER_FAILURE is not set?

Reply:
to get the key, we have to abstract from raw pointers by ptr_map